code/povelikin-rostislav/src/test/java/ru/unn/agile/redblacktree/RedBlackTreeTest.java

```diff
@@ -0,0 +1,230 @@
+package ru.unn.agile.redblacktree;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.*;
+
+import static junit.framework.TestCase.*;
+
+public class RedBlackTreeTest {
+    private static RedBlackTree tree;
+    private static Random random = new Random();
+
+    public RedBlackTreeTest() {
+    }
+
+    @Before
+    public void setUp() {
+        tree = new RedBlackTree();
+    }
+
+    @Test
+    public void defaultConstructedIsEmpty() {
+        assertTrue(tree.isEmpty());
+    }
+
+    @Test
+    public void insertionChangesEmptiness() {
+        tree.insert(8);
+
+        assertFalse(tree.isEmpty());
+    }
+
+    @Test
+    public void emptiedRedBlackTreeIsEmpty() {
+        int expected = 42;
+        tree.insert(expected);
+        tree.remove(expected);
+
+        assertTrue(tree.isEmpty());
+    }
+
+    @Test
+    public void emptyTreeSizeIsZero() {
+        assertEquals(0, tree.getSize());
+    }
+
+    @Test
+    public void notEmptyTreeSizeIsNotZero() {
+        tree.insert(42);
+        assertFalse(tree.getSize() == 0);
+    }
+
+    @Test
+    public void insertionChangesSize() {
+        int threshold = 8;
+        int counter = 0;
+        while (counter < threshold) {
+            ++counter;
+            final int newSize = tree.getSize() + 1;
+            tree.insert(counter);
+            assertEquals(tree.getSize(), newSize);
+        }
+    }
+
+    @Test
+    public void canFindInserted() {
+        int expected = 42;
+        tree.insert(expected);
+
+        assertTrue(tree.find(expected));
+    }
+
+    @Test
+    public void insertCornerCaseMax() {
+        tree.insert(Integer.MAX_VALUE);
+
+        assertTrue(tree.find(Integer.MAX_VALUE));
+    }
+
+    @Test
+    public void insertCornerCaseMin() {
+        tree.insert(Integer.MIN_VALUE);
+
+        assertTrue(tree.find(Integer.MIN_VALUE));
+    }
+
+    @Test
+    public void removeCornerCaseMin() {
+        tree.insert(Integer.MIN_VALUE);
+        tree.remove(Integer.MIN_VALUE);
+
+        assertFalse(tree.find(Integer.MIN_VALUE));
+    }
+
+    @Test
+    public void removeCornerCaseMax() {
+        tree.insert(Integer.MAX_VALUE);
+        tree.remove(Integer.MAX_VALUE);
+
+        assertFalse(tree.find(Integer.MAX_VALUE));
+    }
+
+    @Test
+    public void insertSameNumberIncrementsSize() {
+        int number = 42;
+        tree.insert(number);
+        tree.insert(number);
+        assertEquals(2, tree.getSize());
+    }
+
+    @Test
+    public void insertRandomNumberSequence() {
+        for (int counter = 0; counter < 1000000; ++counter) {
```

Comment:
What's the reson for doing so many iterations? I know it doesn't affect speed, but I'm just curious.

Reply:
That's a kind of stress unit test since it's a random number sequence. The more we add - the more stress we put on a red black tree